### ISSUE TITLE
Please allow for restic to be used as a "Go library"

### ISSUE THREAD

**Issue by thewilli** [restic/restic#4406]
Output of `restic version`
--------------------------

```restic 0.15.2 compiled with go1.20.3 on darwin/arm64```

What should restic do differently? Which functionality do you think we should add?
----------------------------------------------------------------------------------

Restic does not only do a good job as a backup utility, but is also written in Go just as many cloud-native and Kubernetes-based tools / operators are. Allowing restic to be used as part of another application instead of spawning a Go binary from another Go binary would not only be more resource-efficient, but would also allow to have a better control about the state and behavior.

Restic has most of its logic as an [`internal`](https://github.com/restic/restic/tree/e990d3d/internal) package, which [prevents](https://docs.google.com/document/d/1e8kOo3r51b2BWtTs_1uADIA5djfXhPT36s6eHVRIvaU/edit?pli=1) other libraries from importing. 

Please be less _restrictive_ when it comes to declaring parts of restic `internal` to allow other projects to use restic as a library.

What are you trying to do? What problem would this solve?
---------------------------------------------------------

I'd like to use restic as part of a larger construct (Kubernetes Operator) to automate different DR scenarios. I'd like to provide detailed information about the status, which could be easily achieved by a custom [StatusReporter](https://github.com/restic/restic/blob/e990d3d/internal/ui/backup/progress.go#L14-L25) if this wasn't _internal_.

Other use-cases exist (see for example [restic-robot](https://github.com/Southclaws/restic-robot)), where a Go application spawns restic as another process, introducing overhead and the binary as additional dependency.

Adding more flexibility and use-cases would increase the potential for people to contribute back to restic and increase its overall popularity.

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------

The flexibility to use restic even on Windows due to its Go nature is just perfect and makes a big difference when comparing it with other utilities like borg.

**Comment by MichaelEischer** [comment]
The problem with exposing the internals of restic is that this would cause a lot of internal interfaces to essentially become an official API. Quite a few of these APIs expect to be called in the correct way to be safe. It is currently also not feasible to give _any_ guarantees whatsoever regarding stability of these interfaces. Providing such guarantees even for some of these interfaces would require a massive refactoring as preparation. The `cmd/restic` package currently also contains a significant amount of logic, which would have to be refactored first.

I understand the appeal of an API, however, I currently don't see how we could get there anytime soon with a reasonable amount of work.

**Comment by LawyZheng** [comment]
Any updates on this issue?
I agree with the point that many packages are `internal` due to the stability consideration.
But is it possible to refactor `cmd/restic` as a package not as a `main` package?
If it is not in the `main` package, it is possible to call with `go:linkname` as a hacking method.
I think this could be refactored easier and sooner at this moment.

**Comment by MichaelEischer** [comment]
> Would you, in general, accept a first pull request targeting the movement of some of the logic in `cmd/restic` into a new `pkg/restic` package?

I'm rather skeptical about that. The only thing in there that's remotely stable is the interface of the `run*` functions. However, these _expect_ to be run as a separate program. At that level of abstraction it would be easier (more maintainable, ...) to just call restic as external binary.

And everything that's lower level would immediately require a significant amount of interfaces from interfaces from `internal/` to be of any use. A low-level abstraction would also mean that high-level stuff like "backup" aren't readily available.

Can you be a bit more precise on what you have in mind?

**Comment by thewilli** [comment]
> Can you be a bit more precise on what you have in mind?

Sure. I'd consider a short-term and a long-term refactoring.

### Short-term refactoring

Most users interested in this topic will probably be just interested in *using* restic from withing Go, instead of *extending* or *modifying* it. As some kind of *low-hanging fruit*, I'd therefore propose to create a simplified interface (to be understood as *API*, not * Go interface*) to the functionality currently provided by the `cmd/restic` package: Each of the CLI (sub-) commands should be invokable from Go, so the full error handling and runtime control part (e.g. context-passing) is controllable from outside. This follows the idea of the [Go standard project layout](https://github.com/golang-standards/project-layout#go-directories) for the `/cmd` packages:

> It's common to have a small main function that imports and invokes the code from the /internal and /pkg directories and nothing else.

Long story short: I'd create a small abstraction layer for the CLI invocation that contains all the current logic of the `cmd` package to be used by both, the `cmd/restic` package and some external programmers. It would not allow to extend restic in any way (except for runtime aspects like custom [ProgressPrinter](https://github.com/restic/restic/blob/e990d3d/internal/ui/backup/progress.go#L14-L25)), but become a starting point from using restic in existing Go-based applications and services.

This approach has 2 advantages imo:

1. We get a way faster result than refactoring large parts of the restic codebase
2. We get a simple interface for the future, to be used by those who *just* require the basic restic functionality in their applications

### Long-term refactoring

restic exposes a stable interface around the low-level constructs like the [Repository](https://github.com/restic/restic/blob/eef0ee7/internal/repository/repository.go#L38-L59) struct. This allows not only only to control restic, but also to extend it in a dynamic way - for example to allow backing up non-filesystem sources or extending the supported [backend types](https://github.com/restic/restic/blob/f490288/internal/restic/backend.go#L16-L71). 
This will - as you already stated out - require lots of work, and will continue to be an ongoing process for some time.



@MichaelEischer What do you think of this approach?

**Comment by MichaelEischer** [comment]
> This follows the idea of the [Go standard project layout](https://github.com/golang-standards/project-layout#go-directories) for the `/cmd` packages:

The line before that states:

> If the code is not reusable or if you don't want others to reuse it, put that code in the /internal directory.

> I'd therefore propose to create a simplified interface (to be understood as _API_, not * Go interface*) to the functionality currently provided by the `cmd/restic` package: Each of the CLI (sub-) commands should be invokable from Go, so the full error handling and runtime control part (e.g. context-passing) is controllable from outside.

Is "simplified interface" intended to mean "a wrapper around the `run*` functions"? There are over 30 (!) such `run*` functions each accepting several unique options. That means a massive amount of wrapper code (even a pretty trivial wrapper would amount to 1k+ lines) which would have to be maintained. I don't have time for that; we're already barely able to keep the pile of PRs from growing.

There are also several parts of restic that work with global state such as the signal handlers used to handle SIGINT and SIGHUP or the way the lock refresh currently works. And don't forget the log output...

Neither wrapper function could provide a more structured output than right now, unless we start with a massive refactoring. All commands currently output text (or JSON if you're lucky) to stdout/stderr. In case of an error, restic exits with a non-zero exit code and prints a corresponding error message. And to cancel a restic operation just send it a SIGINT signal. Thus, the difference in functionality between the current CLI interface and a Go interface would be next to zero. The main difference would be that we as project end up having to maintain a Go interface that is actually a badly concealed CLI interface. Even worse, in contrast to independent restic child processes that interface won't be able to run multiple restic commands in parallel (that is, unless we massively refactor the code, once again).

**So I really don't get what benefit this should provide. All I see is several months of work without any notable benefit.**

> (except for runtime aspects like custom [ProgressPrinter](https://github.com/restic/restic/blob/e990d3d/internal/ui/backup/progress.go#L14-L25))

Huh? That interface is tied to several internal data structures and is for sure _not_ stable. It would also require some way to inject it.

**Comment by thewilli** [comment]
@MichaelEischer I do see your points. As for the short-term, I do think to provide an interface to (most) CLI functions would provide a benefit for many, even if this means some abstractions / boilerplate code. Being able to control restic, maybe injecting a custom Context to abort an operation, would help to include it in other tools.

The whole discussion is just that I don't have too much time, either. I'm willing to start this, but only if you think that it has a real chance to get merged.

**Comment by konidev20** [comment]
@MichaelEischer, for the short term, can restic provide a library for working with the repository? 

Example of functionality:
1. Walk a snapshot: list files/directories
2. Find a file in the repository
3. Calculate the amount of Data deduplication
4. Calculate the amount of Index memory usage

Again, this is like asking for wrappers around snapshots, ls, and cat commands. But I think we will get some basic APIs to work with a restic repository.

**Comment by Enrico204** [comment]
> @MichaelEischer, for the short term, can restic provide a library for working with the repository?
> 
> Example of functionality:
> 
>     1. Walk a snapshot: list files/directories
> 
>     2. Find a file in the repository
> 
>     3. Calculate the amount of Data deduplication
> 
>     4. Calculate the amount of Index memory usage
> 
> 
> Again, this is like asking for wrappers around snapshots, ls, and cat commands. But I think we will get some basic APIs to work with a restic repository.

You can use the `restic` command and the JSON output format to get this information programmatically.

**Comment by konidev20** [comment]
> > @MichaelEischer, for the short term, can restic provide a library for working with the repository?
> 
> > 
> 
> > Example of functionality:
> 
> > 
> 
> >     1. Walk a snapshot: list files/directories
> 
> > 
> 
> >     2. Find a file in the repository
> 
> > 
> 
> >     3. Calculate the amount of Data deduplication
> 
> > 
> 
> >     4. Calculate the amount of Index memory usage
> 
> > 
> 
> > 
> 
> > Again, this is like asking for wrappers around snapshots, ls, and cat commands. But I think we will get some basic APIs to work with a restic repository.
> 
> 
> 
> You can use the `restic` command and the JSON output format to get this information programmatically.
> 

This adds an overhead of add the restic binary as a dependency. 

Also, quoting from the original issue: 
> Allowing restic to be used as part of another application instead of spawning a Go binary from another Go binary would not only be more resource-efficient, but would also allow to have a better control about the state and behavior.

**Comment by Enrico204** [comment]
> This adds an overhead of add the restic binary as a dependency.

But it allows supporting only one interface (the CLI) as stable interface. The maintainers' bandwidth is limited.

What is the use case where adding the `restic` binary as dependency would be problematic?

**Comment by konidev20** [comment]
> The maintainers' bandwidth is limited.
Yes, agree. No debate on that.

> What is the use case where adding the restic binary as a dependency would be problematic?
It's not problematic. It's an overhead.

From my experience, let's take the example of browsing a snapshot or multiple snapshots. Using the standard CLI interface with JSON is time-consuming and slow. This is because, every time we run a restic CLI command, it performs the following steps
1. Open the repository
2. Loads the index
3. List snapshots
4. Walk snapshot level
5. For each level of the tree, we would have to repeat steps 1 to 3, till we find our restic node

Now, if we have a simpler interface for the restic internals, we can do steps 1 to 3 and keep indexes, etc. ready for the intended operation. 

As the number of snapshots increases in the repository and the indexes too, this operation becomes a more costly overhead. Do note, that even if the indexes are cached locally, they still have to be decrypted, decompressed, and deserialized. Also, consider scenarios, where caching is disabled, the indexes will have to be downloaded from a different location every time.

The only other proposal I have is to run the restic command as a server only for read operations, where commands like cat, snapshots, find etc. can be implemented and maintained simply. 

I have seen some success in my personal experience. It's not too much overhead, I can share what I have done.

**Comment by Enrico204** [comment]
I see, you are right about the unnecessary steps.

I think that your second suggestion requires less work for the current status of the project, as it might be implemented as a dedicated command. Maybe together with #60 ? (e.g., a REST API that can be used directly or via an embedded web application). This interface will be available also to non-Go programs (in contrast to providing a Go library).

(don't get me wrong, I agree that the cleanest approach is to provide a Go library (and maybe a REST server over it). However, I don't see how to do that, except by doing a very slow and big refactoring)

**Comment by konidev20** [comment]
> This interface will be available also to non-Go programs (in contrast to providing a Go library).

This would be a good start. I will check out gh-60 and see if I can contribute in any way.

 > (don't get me wrong, I agree that the cleanest approach is to provide a Go library (and maybe a REST server over it). However, I don't see how to do that, except by doing a very slow and big refactoring) 

Completely agree that it will take a long time. IMO it will provide a good base for long-term application use of restic. (it may also encourage new developers like myself to step forward and help maintain the project)

**Comment by konidev20** [comment]
@Enrico204 

Found this in the pull-requests

https://github.com/restic/restic/pull/4276

@ducalex  has already implemented something we discussed in this thread.

**Comment by aawsome** [comment]
Just wanted to mention that if you don't rely on Go, there is [rustic_core](https://crates.io/crates/rustic_core) written in Rust. With a bit of work, it should be even possible to write a Go wrapper around that library, see e.g. https://words.filippo.io/rustgo/

**Comment by MichaelEischer** [comment]
@hellodword I don't understand at all what you're trying to achieve. Either way, that doesn't belong into this issue, either create a new issue or a forum post.

**Comment by MichaelEischer** [comment]
More on topic for this issue:

Would there be interest if we provided an **UNSTABLE** library interface for restic? (I'm serious about the unstable part, there will be incompatible changes in every restic version for the foreseeable future. So don't complain if your code breaks) It would at first be limited to several low-level interfaces like the backend and repository. I'm not sure yet whether the first iteration would provide write access or just read access.

**Comment by darkdragon-001** [comment]
> More on topic for this issue:
> 
> Would there be interest if we provided an **UNSTABLE** library interface for restic? (I'm serious about the unstable part, there will be incompatible changes in every restic version for the foreseeable future. So don't complain if your code breaks) It would at first be limited to several low-level interfaces like the backend and repository. I'm not sure yet whether the first iteration would provide write access or just read access.

Why starting with low-level interfaces?
Shouldn't high-level interfaces (basically the cli commands with args passed as variables instead of strings) break less often? Or even more important, have the same breakage semantics as restic itself (major version bump if CLI interface changes in an incompatible way)?

**Comment by MichaelEischer** [comment]
Making the high-level interfaces available would require moving most code out of `cmd/restic`. And that unfortunately requires a lot of work (there are currently 16k lines of code in that folder, that might have to be touched).

There's also the question what a "high-level interface" is. Is it just a different way to call the CLI command or do we need a complete new set of interfaces? For example, prompting the user for a password doesn't make much sense when restic is used as library; but that could currently happen. And what about the CLI output? That definitely won't work for a library, so we'll need to restructure at least some of the code. (Although, one could argue that you can redirect stdout and just parse the output there, but then you can just as well call restic externally)

**Comment by konidev20** [comment]
Can we start off with a simple HTTP API approach to provide repository read capabilities for common tasks; including listing snapshots, listing files (with some batching), listing tree and data blobs, find command etc. 

Use a command like `restic explore` to start a simple HTTP server, which will load the index once and will have an API or an auto-refresh timeout, where it will find the missing indexes from memory and will load it. I've implemented this for one of my internal projects, but would be happy to share it with the community.

I've seen faster load times for things like listing snapshots and other simple read operations on the restic repository. The main issue I intended to solve with the simple HTTP server is the following point: 

> From my experience, let's take the example of browsing a snapshot or multiple snapshots. Using the standard CLI interface with JSON is time-consuming and slow. This is because, every time we run a restic CLI command, it performs the following steps
> 1. Open the repository
> 2. Loads the index
> 3. List snapshots
> 4. Walk snapshot level
> 5. For each level of the tree, we would have to repeat steps 1 to 3, till we find our restic node

Scripting operations for read operations on the repo have a significant overhead. 

Example of proposed APIs:
LIST all snapshots
```
GET  /api/v1/repo/snapshots
Headers: X-RESTIC-PASSWORD: <repo-password>
Request Body: NA
Response Body:
[
  {
    "snapshotID": "ced56sds....",
  },
  {
    "snapshotID": "ad56sds....",
  }
]
```

LIST files in a snapshot
```
GET  /api/v1/repo/snapshot/files
Headers: X-RESTIC-PASSWORD: <repo-password>
Request Body:
{
"snapshotID":"<snapshot-id>",
"path":"path-to-filter:optional",
"skip": "<integer>",
"take": "<integer>"
}
Response Body:
[
{
"path": "/E/folder-1",
"name": "folder-1",
"mtime": "2023-02-21T19:24:58.4628969+05:30",
"type": "dir"
},
{
"path": "/E/folder-1/file-1",
"name": "file-1",
"mtime": "2023-02-21T19:24:58.4628969+05:30",
"type": "file"
}
]
```

**Comment by MichaelEischer** [comment]
> Can we start off with a simple HTTP API approach to provide repository read capabilities for common tasks; including listing snapshots, listing files (with some batching), listing tree and data blobs, find command etc.

That looks rather similar to https://github.com/restic/restic/pull/4276 . For the use case you've cited, either the just mentioned PR or a WebDAV interface are the way to go.

Anything more than those interfaces adds yet another massive API surface to restic, which complicates everything. The previously mentioned refactoring of `cmd/restic` would still be necessary to not duplicate large amounts of code. Commands like `find` are also not particularly well suited for an HTTP API, as they are very long running.

**Comment by prologic** [comment]
Hey team 👋  Where do we stand on this issue? I've been using restic for years, quite successfully via the standalone cli `restic` for backups (_pbviously_), but one thing missing is a Go library that can be imported and embedded into another application, avoiding having to shell out to `restic` and read/write from a pipe, etc.

There's been a lot of discussion in this issue over the past 3 years, and I can see it's been labelled as "Needs Direction'.

Any updates / thoughts on this? Is there enough of a demand to make this happen? I certainly would like to see this happen as it makes my own integration story much clener/nicer.
